src/payloads.rs

```diff
@@ -0,0 +1,245 @@
+        .with_to(MULTICALL3_ADDRESS)
+        .with_input(aggregate3_value_payload)
+        .with_value(total);
+    Ok(tx)
+}
+
+/// Predict safe address deployed by the edge node
+pub fn edge_node_predict_safe_address(
+    hopr_node_stake_factory_address: Address,
+    hopr_channels_address: Address,
+    nonce: U256,
+    admins: Vec<Address>,
+) -> Result<Address, HelperErrors> {
+    if admins.is_empty() {
+        return Err(HelperErrors::MissingParameter(
+            "At least one admin address must be provided".into(),
+        ));
+    }
+
+    // build a new temporary admin
+    let mut temporary_admins: Vec<Address> = admins.clone();
+    temporary_admins.insert(0, MULTICALL3_ADDRESS);
+    info!(
+        "temporary_admins expands from admin from {:?} addresses to {:?}",
+        admins.len(),
+        temporary_admins.len()
+    );
+
+    // build the default permissions of capabilities
+    let default_target = U256::from_str(format!("{hopr_channels_address:?}{DEFAULT_CAPABILITY_PERMISSIONS}").as_str())
+        .map_err(|e| HelperErrors::ParseError(format!("Invalid default_target format: {e}")))?;
+    debug!("default target {:?}", default_target);
+
+    let safe_address = predict_safe_address(
+        hopr_node_stake_factory_address,
+        temporary_admins.clone(),
+        nonce.into(),
+        SAFE_COMPATIBILITYFALLBACKHANDLER_ADDRESS,
+        SAFE_SAFE_L2_ADDRESS,
+        SAFE_SAFEPROXYFACTORY_ADDRESS,
+    )?;
+    Ok(safe_address)
+}
+
+pub fn edge_node_predict_module_address(
+    hopr_channels_address: Address,
+    predicted_safe_address: Address,
+    nonce: U256,
+) -> Result<Vec<u8>, HelperErrors> {
+    // build the default permissions of capabilities
+    let default_target = U256::from_str(format!("{hopr_channels_address:?}{DEFAULT_CAPABILITY_PERMISSIONS}").as_str())
+        .map_err(|e| HelperErrors::ParseError(format!("Invalid default_target format: {e}")))?;
+    debug!("default target {:?}", default_target);
+
+    let predict_module_address_payload = predictModuleAddress_1Call {
+        caller: MULTICALL3_ADDRESS,
+        nonce,
+        safe: predicted_safe_address,
+        defaultTarget: default_target.into(),
+    }
+    .abi_encode();
+    Ok(predict_module_address_payload)
+}
+
+/// Payload for deploying safe module with target addresses and node addresses
+/// This payload assumes the threshold is one.
+/// The prediceted safe address can be obtained by calling `edge_node_predict_safe_address`
+/// The predicted module address can be obtained by calling `predictModuleAddress_1` on `hopr_node_stake_factory`
+/// providing MULTICALL3_ADDRESS, nonce.into(), safe_address, default_target
+pub fn edge_node_deploy_safe_module_with_targets_and_nodes_payload(
+    predicted_safe_address: Address,
+    predicted_module_address: Address,
+    hopr_node_stake_factory_address: Address,
+    hopr_channels_address: Address,
+    node_addresses: Option<Vec<Address>>,
```

Comment:
Would an empty Vec be a sufficient indication of no Addresses?

Reply:
In terms of what the code does it would be the same.

Reply:
For this function, no. We have to have predicted safe address